### PR TITLE
Mnaveau/use shared pointer

### PR DESCRIPTION
**Pull request by MaximilienNaveau**
[//]: # "Thanks for your contribution.  To make life of the reviewers easier,"
[//]: # "please give this pull request a meaningful title and provide the"
[//]: # "requested information below."

## Description

[//]: # "Description of what you did.  If it is more complex, consider to add a"
[//]: # "'Summary' section."

Modify the API to use std::shared_ptr. This should prevent potential memory missuses.

## How I Tested

[//]: # "Explain how you tested your changes"

I import the python bindings and tested the addresses.
TODO test on a real robot!

## I fulfilled the following requirements

[//]: # "Please make sure you followed these steps before requesting a review."
[//]: # "Check the boxes in the list below, when done."

- [x] All new code is formatted according to our style guide (for C++ run clang-format, for Python, run flake8 and fix all warnings).
- [x] All new functions/classes are documented and existing documentation is updated according to changes.
- [x] No commented code from testing/debugging is kept (unless there is a good reason to keep it).

### CODE REVIEW

**Comment by MaximilienNaveau** [comment]
@Lhumd and @jviereck can I ask you guys to test the python bindings on solo12?
There is a `demos/demo_solo12` file implemented by @jviereck that should work.

Can you let me know?

**Comment by jviereck** [comment]
Thanks for your changes @MaximilienNaveau .

After compiling and sourcing the workspace, I am getting this error:

```bash
$ ipython
In [1]: run demo_solo12.py
if_name: enp5s0f1
Using Ethernet (enp5s0f1)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/dev/blmc_testing/workspace/src/odri_control_interface/demos/demo_solo12.py in <module>
     23 # Read the values once here. The returned values are views on the data and
     24 # update after the call to `robot.parse_sensor_data()`.
---> 25 imu_attitude = robot.imu.attitude_euler
     26 positions = robot.joints.positions
     27 velocities = robot.joints.velocities

TypeError: No Python class registered for C++ class std::shared_ptr<odri_control_interface::IMU>

```

**Comment by MaximilienNaveau** [comment]
Ok I know what the problem is

**Comment by MaximilienNaveau** [comment]
This commit should fix the issue, sorry that was a dumb mistake.
Let me know when you have the chance to check this back

**Comment by MaximilienNaveau** [comment]
@jviereck @Lhumd this is finally mature, can you take look at the code an see if I didn't do something too crazy?
Thanks

**Comment by Lhumd** [comment]
> @jviereck @Lhumd this is finally mature, can you take look at the code an see if I didn't do something too crazy?
> Thanks

There is a problem with reading IMU data and I think it comes from here. It would be better to have a successful test on Bolt and then merge it.

File: include/odri_control_interface/joint_modules.hpp [hunk 1]
```diff
@@ -64,28 +63,26 @@ class JointModules
...
-        double gear_ratios,
-        double max_currents,
-        RefVectorXb reverse_polarities,
-        RefVectorXd lower_joint_limits,
-        RefVectorXd upper_joint_limits,
-        double max_joint_velocities,
-        double safety_damping
-    );
+    JointModules(const std::shared_ptr<MasterBoardInterface>& robot_if,
+                 ConstRefVectorXi motor_numbers,
```
Comment (jviereck) [inline include/odri_control_interface/joint_modules.hpp:67]: Did you check if the python bindings work? I had troubles when I wanted to use the integers before as it looks like the python binding was expecting long as type. That's why I implemented the motor_numbers as `RefVectorX[l]` before.

**Review by jviereck — COMMENTED** [review]
Thanks for the changes!

Did you check if the code works with the python bindings/python examples? Should I try it on the real robot?

File: include/odri_control_interface/joint_modules.hpp (hunk 1 above)
Comment (MaximilienNaveau) [reply include/odri_control_interface/joint_modules.hpp:67]: Ok you should try it and let me know if the python bindings are working or not.
If not I will provide a setter/accessor for `VectorXl` in the python bindings

Comment (MaximilienNaveau) [reply include/odri_control_interface/joint_modules.hpp:67]: The yaml parsing goes well and the objects are returned properly in python at least.
I am checking the construction of the robot by hand.

**Comment by jviereck** [comment]
Trying the latest branch, I am getting this error now:

```
In [1]: run demo_solo12_from_yaml.py
if_name: enp5s0f1
Using Ethernet (enp5s0f1)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
~/dev/blmc_testing/workspace/src/odri_control_interface/demos/demo_solo12_from_yaml.py in <module>
     32 # update after the call to `robot.parse_sensor_data()`.
     33 imu_attitude = robot.imu.attitude_euler
---> 34 positions = robot.joints.positions
     35 velocities = robot.joints.velocities
     36 

TypeError: No to_python (by-value) converter found for C++ type: Eigen::Ref<Eigen::Matrix<double, -1, 1, 0, -1, 1> const, 0, Eigen::InnerStride<1> >
```

**Comment by MaximilienNaveau** [comment]
@jviereck can you check again?

**Comment by jviereck** [comment]
With the above change (see last commit from me) I am able to run the python example on the real robot!

**Comment by MaximilienNaveau** [comment]
Cool, python bindings are working, hence we need to figure out why the imu is buggy on bolt.

**Comment by jviereck** [comment]
> hence we need to figure out why the imu is buggy on bolt.

Shall this PR be merged already or do you want to figure out what's the problem on bolt first?

**Comment by MaximilienNaveau** [comment]
I want to know if the imu behavior comes from the `master_board_sdk` or the odri, investigation happens here:
https://github.com/open-dynamic-robot-initiative/bolt/issues/5

**Comment by MaximilienNaveau** [comment]
Things work fine on bolt Debug and Release!
So I think we should merge and fix further bugs with additional PRs.

**Pull request merged**